packages/here-wallet/src/lib/utils.ts

```diff
@@ -0,0 +1,105 @@
+import { Network, Optional, Transaction } from "@near-wallet-selector/core";
+import { createAction } from "@near-wallet-selector/wallet-utils";
+import * as BN from "bn.js";
+import {
+  utils,
+  connect,
+  keyStores,
+  WalletConnection,
+  transactions as nearTransactions,
+} from "near-api-js";
+
+export interface HereWalletState {
+  wallet: WalletConnection;
+  keyStore: keyStores.BrowserLocalStorageKeyStore;
+}
+
+export interface HereConfiguration {
+  hereWallet: string;
+  hereContract: string;
+}
+
+export const hereConfigurations: Record<string, HereConfiguration> = {
+  mainnet: {
+    hereWallet: "https://web.herewallet.app",
+    hereContract: "storage.herewallet.near",
+  },
+  testnet: {
+    hereWallet: "https://web.testnet.herewallet.app",
+    hereContract: "storage.herewallet.testnet",
+  },
+};
+
+const setupWalletState = async (
+  config: HereConfiguration,
+  network: Network
+): Promise<HereWalletState> => {
+  const keyStore = new keyStores.BrowserLocalStorageKeyStore();
+  const near = await connect({
+    keyStore,
+    walletUrl: config.hereWallet,
+    headers: {},
+    ...network,
+  });
+
+  const wallet = new WalletConnection(near, "here_app");
+
+  // Cleanup up any pending keys (cancelled logins).
+  if (!wallet.isSignedIn()) {
+    await keyStore.clear();
+  }
+
+  return { wallet, keyStore };
+};
+
+const getHereBalance = async (
+  state: HereWalletState,
+  config: HereConfiguration
+): Promise<BN> => {
+  const params = { account_id: state.wallet.getAccountId() };
+  const hereCoins = await state.wallet
+    .account()
+    .viewFunction(config.hereContract, "ft_balance_of", params)
+    .catch(() => "0");
+
+  return new BN(hereCoins);
+};
+
+const transformTransactions = async (
+  state: HereWalletState,
+  transactions: Array<Optional<Transaction, "signerId">>
+) => {
+  const account = state.wallet.account();
+  const { networkId, signer, provider } = account.connection;
+  const localKey = await signer.getPublicKey(account.accountId, networkId);
+
+  return Promise.all(
```

Comment:
Can we avoid using a `Promise.all()` here?

This is to potentially avoid an issue that was reported here in Wallet Selector in the past: this avoids nonce increment issues and a concept called “transaction dependency relationship“ where the order of the transaction matters.

You can use a regular for loop to create these transactions in series.

Reply:
But Promise.all doesn't change the order of the array elements:
https://stackoverflow.com/questions/28066429/promise-all-order-of-resolved-values

My-near-wallet has a similar code in the package.